In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

df = pd.read_csv("world_bank_data_2025.csv")
df['Log_GDP_per_Capita'] = np.log(df['GDP per Capita (Current USD)'])

features = [
    'Interest Rate (Real, %)',
    'Government Expense (% of GDP)',
    'Tax Revenue (% of GDP)',
    'Gross National Income (USD)',
    'GDP per Capita (Current USD)',
]

clean_df = df[features].dropna().copy()

clean_df['LogPerCapita'] = np.log(clean_df['GDP per Capita (Current USD)'])

X_features = [
    'Interest Rate (Real, %)',
    'Government Expense (% of GDP)',
    'Gross National Income (USD)',
    'Tax Revenue (% of GDP)',
    'GDP per Capita (Current USD)'
]

x = clean_df[X_features]
y = clean_df['LogPerCapita']

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42
)

print('Training features shape:', X_train.shape)
print('Testing features shape:', X_test.shape)

Training features shape: (900, 5)
Testing features shape: (225, 5)


#  Pick an initial set of features for X and the target feature y.  Explain why you made this choice.
My target feature is LogPerCapita (log of GDP per Capita - purely to aid in correlation detection, can easily be converted back to GDP per Capita).

X features are Interest Rate, Government Expense, Tax Revenue, Inflation, and Unemployement Rate. 

I chose the target feature and am using log because there is a wide range of GDP's from over 200 countries around the world and log will help linearize the data without hurting the values of GDP per capita themselves. I think this will aid in the model's prediction of GDP per capita given the features lsited prior. I chose the x features because these were shown to be among the highest potential/correlation alongside Log GDP per Capita. They also give a good range of information about the state of a country's economy.

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

y_pred = lr_model.predict(X_test)

# Compute evaluation metrics
r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)

print('--- Linear Regression Performance ---')
print(f'R2 Score: {r2:.4f}')
print(f'Mean Squared Error (MSE): {mse:.4f}')
print(f'Root Mean Squared Error (RMSE): {rmse:.4f}')
print(f'Mean Absolute Error (MAE): {mae:.4f}')

--- Linear Regression Performance ---
R2 Score: 0.0715
Mean Squared Error (MSE): 1.7923
Root Mean Squared Error (RMSE): 1.3388
Mean Absolute Error (MAE): 1.0519


# Comment on the results
My r2 score means that only about 7.2% of my variance in LogPerCapita is explained by my model. That's really not good. The MAE score is also pretty bad because while 1.05 units off during my model's predictions may seem small, because the Per Capita is logged, that 1.05 is actually a lot more influential than it would seem. My MSE and RMSE are also a little alarming because if my RMSE is larger than my MAE, that means there are outliers affecting the performance of my model.

# Note:
Back when I did my intial exploration, I used GPD per Capita and Gross National Income. I've changed it because the difference in how they were measured (USD vs. % of GDP) was throwing my model's performance out the window 

In [ ]:
features = [
    'Interest Rate (Real, %)',
    'Government Expense (% of GDP)',
    'Tax Revenue (% of GDP)',
    'Inflation (CPI %)',
    'Unemployment Rate (%)',
    'GDP per Capita (Current USD)',
]

clean_df = df[features].dropna().copy()

clean_df['LogPerCapita'] = np.log(clean_df['GDP per Capita (Current USD)'])

X_features = [
    'Interest Rate (Real, %)',
    'Government Expense (% of GDP)',
    'Tax Revenue (% of GDP)',
    'Inflation (CPI %)',
    'Unemployment Rate (%)',
]

x = clean_df[X_features]
y = clean_df['LogPerCapita']

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)


--- Linear Regression Performance ---
R2 Score: 0.2568
Mean Squared Error (MSE): 1.2589
Root Mean Squared Error (RMSE): 1.1220
Mean Absolute Error (MAE): 0.8636


In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

y_pred = lr_model.predict(X_test)

# Compute evaluation metrics
r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)

print('--- Final Linear Regression Performance ---')
print(f'R2 Score: {r2:.4f}')
print(f'Mean Squared Error (MSE): {mse:.4f}')
print(f'Root Mean Squared Error (RMSE): {rmse:.4f}')
print(f'Mean Absolute Error (MAE): {mae:.4f}')

# Look at the parameters you found and discuss what you have learned.
I learned that it's important to make sure all of your features are operating under the same measurement/template of measurement (i.e. there can be percentages of 2 things but both need to be percentages, rather than numbers and a percentage). Making these adjustments greatly improved the performance of my model and helped explain how much of the variance in data was able to be explained by my model's features. I will say that due to the nature of the topic, there are a lot of reasons and forces behind why the GDP per capita is what it is. My model is observing a few different features and attempting to make predictions based on those few features so the linear regression, while providing a good baseline, will never be fully capable of simulating/predicting what all goes into GDP per Capita. 